webhook.go

```diff
@@ -428,22 +421,37 @@ func (whsvr *WebhookServer) mutateInitialization(pod corev1.Pod, req *v1beta1.Ad
 }
 
 // create mutation patch for resoures
-func createPatch(createVaultCert bool, pod *corev1.Pod, initcontainerConfig *InitContainerConfig) ([]byte, error) {
+func createPatch(pod *corev1.Pod, initcontainerConfig *InitContainerConfig) ([]byte, error) {
 	var patch []patchOperation
-	annotations := initcontainerConfig.Annotations
 
-	// additions
-	patch = append(patch, updateAnnotation(pod.Annotations, annotations)...)
-	if createVaultCert {
-		patch = append(patch, addContainer(pod.Spec.InitContainers, initcontainerConfig.InitContainers, "/spec/initContainers")...)
+	// by default, mutate the pod with the new sidecar and required volumes
+	var mutateAll bool = true
+
+	// check if the pod already mutated, if so, only update the annotations
+	currentAnnotations := pod.ObjectMeta.GetAnnotations()
+	if currentAnnotations != nil {
+		status := currentAnnotations[admissionWebhookAnnotationStatusKey]
+		if strings.ToLower(status) == "mutated" {
+			glog.Infof("Pod already mutated. Update only the annotations")
+			mutateAll = false
+		}
 	}
-	patch = append(patch, addContainer(pod.Spec.Containers, initcontainerConfig.SidecarContainers,
-		"/spec/containers")...)
-	patch = append(patch, addVolume(pod.Spec.Volumes, initcontainerConfig.Volumes, "/spec/volumes")...)
 
-	for i, c := range pod.Spec.Containers {
-		glog.Infof("add vol mounts : %#v", addVolumeMount(c.VolumeMounts, initcontainerConfig.AddVolumeMounts, fmt.Sprintf("/spec/containers/%d/volumeMounts", i)))
-		patch = append(patch, addVolumeMount(c.VolumeMounts, initcontainerConfig.AddVolumeMounts, fmt.Sprintf("/spec/containers/%d/volumeMounts", i))...)
+	// always get updated annotations
+	annotations := initcontainerConfig.Annotations
+	patch = append(patch, updateAnnotation(pod.Annotations, annotations)...)
+	glog.Infof("Annotations are always updated. New values: %#v", annotations)
+
+	// update everything only if not mutated before
+	if mutateAll {
+		patch = append(patch, addContainer(pod.Spec.Containers, initcontainerConfig.SidecarContainers,
```

Comment:
We probably should rename `initcontainerConfig` to reflect what it does now. Maybe like `tsiInjectConfig` or `tsiMutateConfig`.